internal/app/referenceclient/transports.go

```diff
@@ -0,0 +1,180 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package referenceclient
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	conformancev1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1"
+	"connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1/conformancev1connect"
+	"github.com/quic-go/quic-go"
+	"github.com/quic-go/quic-go/http3"
+	"golang.org/x/net/http2"
+)
+
+type transportSpec struct {
+	httpVersion   conformancev1.HTTPVersion
+	serverTLSCert string
+	clientTLSCert string
+	clientTLSKey  string
+}
+
+type transports struct {
+	cache sync.Map // map[transportSpec]http.RoundTripper
+}
+
+func (t *transports) get(req *conformancev1.ClientCompatRequest) (http.RoundTripper, error) {
+	spec := transportSpec{
+		httpVersion:   req.GetHttpVersion(),
+		serverTLSCert: string(req.GetServerTlsCert()),
+		clientTLSCert: string(req.GetClientTlsCreds().GetCert()),
+		clientTLSKey:  string(req.GetClientTlsCreds().GetKey()),
+	}
+
+	// Optimistically skip logic if it's already cached. We will still do an
+	// atomic store to share the transport in all cases even if this misses.
+	if tr, ok := t.cache.Load(spec); ok {
+		return tr.(http.RoundTripper), nil //nolint:errcheck,forcetypeassert
+	}
+
+	tlsConf, err := createTLSConfig(req)
+	if err != nil {
+		return nil, err
+	}
+	var transport http.RoundTripper
+	switch req.HttpVersion {
+	case conformancev1.HTTPVersion_HTTP_VERSION_1:
+		if tlsConf != nil {
+			tlsConf.NextProtos = []string{"http/1.1"}
+		}
+		tx := &http.Transport{
+			DisableCompression: true,
+			TLSClientConfig:    tlsConf,
+		}
+		transport = roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+			resp, err := tx.RoundTrip(req)
+			if resp != nil &&
+				strings.HasSuffix(req.URL.Path, conformancev1connect.ConformanceServiceBidiStreamProcedure) {
+				// To force support for bidirectional RPC over HTTP 1.1 (for half-duplex testing),
+				// we "trick" the client into thinking this is HTTP/2. We have to do this because
+				// otherwise, connect-go refuses to support bidi streams over HTTP 1.1.
+				resp.ProtoMajor, resp.ProtoMinor = 2, 0
+			}
+			return resp, err
+		})
+	case conformancev1.HTTPVersion_HTTP_VERSION_2:
+		if tlsConf != nil {
+			tlsConf.NextProtos = []string{"h2"}
+			transport = &http.Transport{
+				DisableCompression: true,
+				TLSClientConfig:    tlsConf,
+				ForceAttemptHTTP2:  true,
+			}
+		} else {
+			transport = &http2.Transport{
```

Comment:
While you're moving this code around, mind adding a TODO comment about net/http's recent support for H2C?
```
// TODO: Now that go 1.24 is required, we should use net/http's builtin support
//       for H2Cand drop the dependency on golang.org/x/net/http2.
```

Reply:
Went ahead and switched over to the native support